### PR TITLE
add issingular

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
@yingboma do you have ideas for other dispatches that would be good? Should there be a fallback?

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Can you take it from here? It seems you have a bunch of ideas so that might be quicker.

**Comment by YingboMa** [comment]
Done. @ChrisRackauckas could you take a look?

File: src/ArrayInterface.jl [hunk 1]
```diff
@@ -106,6 +106,19 @@ has_sparsestruct(x::Type{<:Bidiagonal}) = true
...
 has_sparsestruct(x::Type{<:SymTridiagonal}) = true
 
+"""
+    issingular(A::AbstractMatrix)
+
+Determine whether a given abstract matrix is singular.
+"""
+issingular(A::Matrix) = !issuccess(lu(A, check=false))
+issingular(A::UniformScaling) = A.λ == 0
+issingular(A::Diagonal) = any(iszero,A.diag)
```
Comment (dkarrasch) [inline src/ArrayInterface.jl]: One could add here
```julia
issingular(B::Bidiagonal) = any(iszero, A.dv)
issingular(S::SymTridiagonal) = diaganyzero(iszero, ldlt(S).data)
issingular(T::Tridiagonal) = !issuccess(lu(A, check=false))
```
And perhaps as a fallback (which fulfills the promise of the docstring)
```julia
issingular(A::AbstractMatrix) = issingular(Matrix(A))
issingular(A::AbstractSparseMatrix) = !issuccess(lu(A, check=false))
```

**Comment by ChrisRackauckas** [comment]
Note I reverted the README change @YingboMa but it's expected that `issingular` is the best way to check for it: if it's ever faster to use `rank`, then the dispatch should be to use `rank`.

**Pull request merged**